File: libraries/node/src/sources/sm/parse.ts
/**
 * Parser for StepMania (.sm/.ssc) simfiles.
 */
import type { SmChart, SmFile, Timing } from "./model.js";
import {
	beatValue,
	expressions,
	pairs,
	parseFloatStrict,
	parseIntStrict,
	tokenize
} from "./msd.js";
import { parseNoteData } from "./notes.js";
import { UrcError } from "../../error.js";

const STEP_LANES: Record<string, number> = {
	"dance-single": 4,
	"dance-double": 8,
	"dance-solo": 6,
	"dance-threepanel": 3,
	"pump-single": 5,
	"pump-halfdouble": 6,
	"pump-double": 10,
	"kb7-single": 7,
	"techno-single4": 4,
	"techno-single5": 5,
	"techno-single8": 8,
	"techno-double4": 8,
	"techno-double5": 10,
	"techno-double8": 16,
	"maniax-single": 4,
	"maniax-double": 8,
	"pnm-five": 5,
	"pnm-nine": 9,
	"para-single": 5,
	"ds3ddx-single": 8,
	"ez2-single": 5,
	"ez2-double": 10,
	"ez2-real": 7,
	"kickbox-human": 4,
	"kickbox-quadarm": 4,
	"kickbox-insect": 6,
	"kickbox-arachnid": 8
};
const STEP_ALIASES: Record<string, string> = {
	"ez2-single-hard": "ez2-single",
	para: "para-single"
};

const TIMING_TAGS = new Set([
	"OFFSET",
	"BPMS",
	"STOPS",
	"FREEZES",
	"DELAYS",
	"WARPS",
	"SCROLLS",
	"FAKES",
	"TIMESIGNATURES"
]);

/**
 * Track count for a steps type; unsupported types are an error.
 */
export function resolveLanes(stepsType: string): number {
	const name = STEP_ALIASES[stepsType] ?? stepsType;
	const lanes = STEP_LANES[name];
	if (lanes === undefined)
		throw new UrcError(
			"unsupported-version",
			1,
			`unsupported steps type: ${stepsType || "(missing)"}`
		);
	return lanes;
}

/**
 * Parses a .sm or .ssc simfile into its source model.
 */
export function parseSm(text: string): SmFile {
	const simfile: SmFile = {
		title: "",
		subtitle: "",
		artist: "",
		credit: "",
		timing: emptyTiming(),
		charts: []
	};
	let chart: SmChart | null = null;

	for (const params of tokenize(text)) {
		const tag = params[0].toUpperCase();
		const value = params.length > 1 ? params[1] : "";

		if (tag === "NOTEDATA") {
			chart = emptyChart();
			continue;
		}
		if (tag === "NOTES" || tag === "NOTES2") {
			if (chart !== null) {
				chart.notes = parseNoteData(value, resolveLanes(chart.stepsType));
				simfile.charts.push(chart);
				chart = null;
			} else if (params.length >= 7) {
				const block = emptyChart();
				block.stepsType = params[1].trim();
				block.description = params[2].trim();
				block.difficulty = params[3].trim();
				block.credit = params[2].trim();
				block.notes = parseNoteData(params[6], resolveLanes(block.stepsType));
				simfile.charts.push(block);
			}
			continue;
		}

		if (chart === null)
			songTag(simfile, tag, value);
		else
			chartTag(simfile, chart, tag, value);
	}

	if (simfile.charts.length === 0)
		throw new UrcError("syntax", 1, "no chart in simfile");
	return simfile;
}

function emptyTiming(): Timing {
	return {
		offset: 0,
		bpms: [],
		stops: [],
		delays: [],
		warps: [],
		scrolls: [],
		timesigs: [],
		fakes: []
	};
}

function emptyChart(): SmChart {
	return {
		stepsType: "",
		description: "",
		difficulty: "",
		chartname: "",
		credit: "",
		timing: null,
		notes: []
	};
}

function songTag(simfile: SmFile, tag: string, value: string): void {
	switch (tag) {
		case "TITLE":
			simfile.title = value;
			break;
		case "SUBTITLE":
			simfile.subtitle = value;
			break;
		case "ARTIST":
			simfile.artist = value;
			break;
		case "CREDIT":
			simfile.credit = value;
			break;
		default:
			timingTag(simfile.timing, tag, value);
	}
}

function chartTag(simfile: SmFile, chart: SmChart, tag: string, value: string): void {
	switch (tag) {
		case "STEPSTYPE":
			chart.stepsType = value.trim();
			break;
		case "DESCRIPTION":
			chart.description = value.trim();
			break;
		case "DIFFICULTY":
			chart.difficulty = value.trim();
			break;
		case "CHARTNAME":
			chart.chartname = value.trim();
			break;
		case "CREDIT":
			chart.credit = value;
			break;
		default:
			if (TIMING_TAGS.has(tag)) {
				if (chart.timing === null)
					chart.timing = { ...emptyTiming(), offset: simfile.timing.offset };
				timingTag(chart.timing, tag, value);
			}
	}
}

function timingTag(timing: Timing, tag: string, value: string): void {
	switch (tag) {
		case "OFFSET":
			timing.offset = parseFloatStrict(value);
			break;
		case "BPMS":
			timing.bpms.push(...pairs(value, true));
			break;
		case "STOPS":
		case "FREEZES":
			timing.stops.push(...pairs(value, true));
			break;
		case "DELAYS":
			timing.delays.push(...pairs(value, true));
			break;
		case "WARPS":
			timing.warps.push(...pairs(value));
			break;
		case "SCROLLS":
			timing.scrolls.push(...pairs(value));
			break;
		case "FAKES":
			for (const entry of pairs(value))
				if (entry[1] > 0)
					timing.fakes.push(entry);
			break;
		case "TIMESIGNATURES":
			for (const parts of expressions(value, 3)) {
				const beat = beatValue(parts[0]);
				const numerator = parseIntStrict(parts[1]);
				const denominator = parseIntStrict(parts[2]);
				if (numerator >= 1 && denominator >= 1 && beat >= 0)
					timing.timesigs.push([beat, numerator, denominator]);
			}
	}
}

File: libraries/rust/src/sources/sm/parse.rs
//! Parser for StepMania (`.sm`/`.ssc`) simfiles.

use std::collections::HashMap;

use crate::error::{Result, UrcError};

use super::model::{SmChart, SmFile, Timing};
use super::msd::{expressions, pairs, parse_beat, parse_float, parse_int, tokenize};
use super::notes::parse_note_data;

fn step_lanes() -> HashMap<&'static str, u32> {
    HashMap::from([
        ("dance-single", 4),
        ("dance-double", 8),
        ("dance-solo", 6),
        ("dance-threepanel", 3),
        ("pump-single", 5),
        ("pump-halfdouble", 6),
        ("pump-double", 10),
        ("kb7-single", 7),
        ("techno-single4", 4),
        ("techno-single5", 5),
        ("techno-single8", 8),
        ("techno-double4", 8),
        ("techno-double5", 10),
        ("techno-double8", 16),
        ("maniax-single", 4),
        ("maniax-double", 8),
        ("pnm-five", 5),
        ("pnm-nine", 9),
        ("para-single", 5),
        ("ds3ddx-single", 8),
        ("ez2-single", 5),
        ("ez2-double", 10),
        ("ez2-real", 7),
        ("kickbox-human", 4),
        ("kickbox-quadarm", 4),
        ("kickbox-insect", 6),
        ("kickbox-arachnid", 8),
    ])
}

/// Track count for a steps type; unsupported types are an error.
pub fn resolve_lanes(steps_type: &str) -> Result<u32> {
    let alias = match steps_type {
        "ez2-single-hard" => "ez2-single",
        "para" => "para-single",
        other => other,
    };
    let shown = if steps_type.is_empty() {
        "(missing)"
    } else {
        steps_type
    };
    step_lanes()
        .get(alias)
        .copied()
        .ok_or_else(|| UrcError::new("unsupported-version", 1, format!("unsupported steps type: {shown}")))
}

/// Parses a `.sm` or `.ssc` simfile into its source model.
pub fn parse_sm(text: &str) -> Result<SmFile> {
    let mut simfile = SmFile::default();
    let mut chart: Option<SmChart> = None;

    for params in tokenize(text) {
        let tag = params[0].to_uppercase();
        let value = params.get(1).cloned().unwrap_or_default();

        if tag == "NOTEDATA" {
            chart = Some(SmChart::default());
            continue;
        }
        if tag == "NOTES" || tag == "NOTES2" {
            if let Some(mut open) = chart.take() {
                let lanes = resolve_lanes(&open.steps_type)?;
                open.notes = parse_note_data(&value, lanes)?;
                simfile.charts.push(open);
            } else if params.len() >= 7 {
                let mut block = SmChart {
                    steps_type: params[1].trim().to_string(),
                    description: params[2].trim().to_string(),
                    difficulty: params[3].trim().to_string(),
                    credit: params[2].trim().to_string(),
                    ..Default::default()
                };
                let lanes = resolve_lanes(&block.steps_type)?;
                block.notes = parse_note_data(&params[6], lanes)?;
                simfile.charts.push(block);
            }
            continue;
        }

        match chart {
            None => song_tag(&mut simfile, &tag, &value)?,
            Some(ref mut open) => chart_tag(&simfile, open, &tag, &value)?,
        }
    }

    if simfile.charts.is_empty() {
        return Err(UrcError::new("syntax", 1, "no chart in simfile"));
    }
    Ok(simfile)
}

fn song_tag(simfile: &mut SmFile, tag: &str, value: &str) -> Result<()> {
    match tag {
        "TITLE" => simfile.title = value.to_string(),
        "SUBTITLE" => simfile.subtitle = value.to_string(),
        "ARTIST" => simfile.artist = value.to_string(),
        "CREDIT" => simfile.credit = value.to_string(),
        _ => timing_tag(&mut simfile.timing, tag, value)?,
    }
    Ok(())
}

fn chart_tag(simfile: &SmFile, chart: &mut SmChart, tag: &str, value: &str) -> Result<()> {
    match tag {
        "STEPSTYPE" => chart.steps_type = value.trim().to_string(),
        "DESCRIPTION" => chart.description = value.trim().to_string(),
        "DIFFICULTY" => chart.difficulty = value.trim().to_string(),
        "CHARTNAME" => chart.chartname = value.trim().to_string(),
        "CREDIT" => chart.credit = value.to_string(),
        "OFFSET" | "BPMS" | "STOPS" | "FREEZES" | "DELAYS" | "WARPS" | "SCROLLS" | "FAKES"
        | "TIMESIGNATURES" => {
            let timing = chart.timing.get_or_insert_with(|| Timing {
                offset: simfile.timing.offset,
                ..Default::default()
            });
            timing_tag(timing, tag, value)?;
        }
        _ => {}
    }
    Ok(())
}

fn timing_tag(timing: &mut Timing, tag: &str, value: &str) -> Result<()> {
    match tag {
        "OFFSET" => timing.offset = parse_float(value)?,
        "BPMS" => timing.bpms.extend(pairs(value, true)?),
        "STOPS" | "FREEZES" => timing.stops.extend(pairs(value, true)?),
        "DELAYS" => timing.delays.extend(pairs(value, true)?),
        "WARPS" => timing.warps.extend(pairs(value, false)?),
        "SCROLLS" => timing.scrolls.extend(pairs(value, false)?),
        "FAKES" => timing
            .fakes
            .extend(pairs(value, false)?.into_iter().filter(|entry| entry.1 > 0.0)),
        "TIMESIGNATURES" => {
            for parts in expressions(value, 3)? {
                let beat = parse_beat(&parts[0])?;
                let numerator = parse_int(&parts[1])?;
                let denominator = parse_int(&parts[2])?;
                if numerator >= 1 && denominator >= 1 && beat >= 0.0 {
                    timing
                        .timesigs
                        .push((beat, numerator as u64, denominator as u64));
                }
            }
        }
        _ => {}
    }
    Ok(())
}

